### ISSUE TITLE
Report only on files in a given directory or matching a given glob

### ISSUE THREAD

**Issue by mvz** [colszowka/simplecov#696]
At [reek](https://github.com/troessner/reek), we're using SimpleCov with RSpec, and have the configuation in `.simplecov`. We use SimpleCov 0.16.1, and my `RUBY_DESCRIPTION` is `ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]`.

Here is our current configuration:

```
SimpleCov.start do
  track_files 'lib/**/*.rb'
  add_filter 'lib/reek/version.rb' # version.rb is loaded too early to test
  add_filter 'lib/reek/cli/options.rb' # tested mostly via integration tests
end

SimpleCov.at_exit do
  SimpleCov.result.format!
  SimpleCov.minimum_coverage 98.9
  SimpleCov.minimum_coverage_by_file 81.4
end
```

We are running into the issue that SimpleCov is reporting coverage for our spec files (in `spec/`). We already have some configuration to make sure all files in `lib/` are reported on, except for two which are filtered out. We *could* add yet another filter to remove `spec/`, but that seems an open-ended process (e.g., if we were to add coverage information for our cucumber scenarios, we would have to exclude the related step definitions as well).

There is an [https://github.com/troessner/reek/pull/1420](open pull request) where we're trying to improve the situation by setting `root`, but I don't think that's the right answer, because it breaks other behavior.

There currently seems no configuration that is essentially the inverse of `add_filter` and makes SimpleCov only report on files matching the given expression, i.e., something like this:

```ruby
only_report_on 'lib/**/*.rb'
```

**Comment by PragTob** [comment]
Hi @mvz thanks for your report and thanks for your work on reek! :green_heart: :tada: :rocket: 

Seems like a great idea. I think we'd have to establish some hieararchy of how these work together then (only_report_on + add_filter = ?). All of these should be better documented anyhow though see #610 

Also sorry a bunch for not responding earlier here, you know time etc. :see_no_evil:

**Comment by sferik** [comment]
Spec files showing up in your coverage report is resolved by 02d6c0c5a4dc6c388ab64e1144b41af0eb634c6a, which now filters them out by default, so the extra `add_filter` lines in reek’s setup are no longer needed.

Your broader proposal—a positive `only_report_on 'lib/**/*.rb'`—isn't implemented but you can use a block filter that negates an include pattern, which should have the same effect:

```ruby
SimpleCov.add_filter { |source_file| !source_file.filename.match?(%r{/lib/}) }
```

I’m going to close this issue, but if you’d still like an allowlist-style API, please open a fresh issue so the design questions (precedence vs. `add_filter`, behavior when both are set, glob vs. Regexp inputs) can be discussed in a focused thread.

Thanks for submitting this issue…and for reek!

**Comment by sferik** [comment]
The positive-allowlist API landed in https://github.com/simplecov-ruby/simplecov/pull/1188 as `cover`, so your example becomes:

```ruby
SimpleCov.start do
  cover "lib/**/*.rb"
end
```

This will ship in the v1.0.0 release but you can test it today by replacing the `simplecov` line in your `Gemfile` with:
 
```ruby
gem "simplecov", git: "https://github.com/simplecov-ruby/simplecov.git", branch: "main", require: false
```

I’d greatly appreciate if you’re able to test this before the v1.0.0 release and report and issues you find.

**Comment by mvz** [comment]
@sferik yes, I have tested this and `cover` does exactly what we want, thanks!
